### DISCUSSION TITLE
Creating Tiled images of different sizes on the same image

### DISCUSSION THREAD

**Discussion by Trailboy** [ImageMagick/ImageMagick#2826] [Digital Image Processing]
Hi.

I was wondering if anybody was able to help me with creating tiles from imagemagick?

I need to convert a single image into a 1800x1200 pixel (6x4 inches) tiled composite print that has 2 of 3x2 (fridge magnets sized basically) and 8 of 1.5x1 inch images (key-rings sized).  

With previous help (thank you everybody), I have managed to create a 6x4 sized image that creates an image that is all 3x2 or all 1.5x1, and that works perfectly:

![Imagemagick](https://user-images.githubusercontent.com/64076927/99147797-a79f4d00-267b-11eb-912e-18f5839ac7a1.JPG)

with the code for 4 of 3x2:

```
magick %1 ^
	-gravity center ^
	-extent "%%[fx:w<h&&w/2>h/3?h/3*2:w]x%%[h]" ^
	-extent "%%[fx:w>h&&w/3>h/2?h/2*3:w]x%%[h]" ^
	-extent "%%[w]x%%[fx:w<h&&w/2<h/3?w/2*3:h]" ^
	-extent "%%[w]x%%[fx:w>=h&&w/3<h/2?w/3*2:h]" ^
	-resize 900x900 ^
	-quality 68 ^
	-bordercolor gray -gravity center -border 1 ^
	-duplicate 1 +append ^
	-duplicate 1 -append ^
	-bordercolor gray -gravity center -border 10 ^
	"%~dp1\%~n1_Magnets.jpg"

```

and the code for 16 of 1.5x1:

```
magick %1 ^
	-gravity center ^
	-extent "%%[fx:w<h&&w/2>h/3?h/3*2:w]x%%[h]" ^
	-extent "%%[fx:w>h&&w/3>h/2?h/2*3:w]x%%[h]" ^
	-extent "%%[w]x%%[fx:w<h&&w/2<h/3?w/2*3:h]" ^
	-extent "%%[w]x%%[fx:w>=h&&w/3<h/2?w/3*2:h]" ^
	-resize 450x450 ^
	-quality 68 ^
	-bordercolor gray -gravity center -border 1 ^
	-duplicate 3 +append ^
	-duplicate 3 -append ^
	-bordercolor gray -gravity center -border 10 ^
	"%~dp1\%~n1_Keyrings.jpg"
```

But I am wasting time and money creating both types for customers.  If they could be combined into one print, then it would save me hugely.  

Can anybody help me with this?  

Ideally, something like this, but I don't mind any layout. 

![Imagemagick-2](https://user-images.githubusercontent.com/64076927/99148040-6c058280-267d-11eb-97e8-a37ba6278846.JPG)

Thank you so much.

I'm afraid that my coding skills leave a lot to be desired.

**Comment by GeeMack** [comment]
With ImageMagick you can isolate the different parts of the operation inside sets of parentheses. This example command makes the arrangement with two magnet size and eight key ring size...

    magick %1 ^
        -gravity center ^
        -extent "%%[fx:w<h&&w/2>h/3?h/3*2:w]x%%[h]" ^
        -extent "%%[fx:w>h&&w/3>h/2?h/2*3:w]x%%[h]" ^
        -extent "%%[w]x%%[fx:w<h&&w/2<h/3?w/2*3:h]" ^
        -extent "%%[w]x%%[fx:w>=h&&w/3<h/2?w/3*2:h]" ^
        -quality 68 ^
        -bordercolor gray ^
        -gravity center ^
         ( ^
           -clone 0 ^
           -resize 900x900 ^
           -shave 1 ^
           -border 1 ^
         ) ^
         ( ^
           -clone 0 ^
           -resize 450x450 ^
           -shave 1 ^
           -border 1 ^
           +duplicate +append ^
           +duplicate -append ^
         ) ^
        -delete 0 ^
        +append ^
        +duplicate ^
        -append ^
        -border 10 ^
           "%~dp1\%~n1_Combo.jpg"

That reads the input image and reshapes it to 2:3. Then inside the first parentheses it makes the magnet size image. Inside the next parentheses it makes four key ring size and appends them into a grid. After the parentheses it deletes the re-shaped input image, and horizontally appends the results from both sets of parentheses, then duplicates that and vertically appends them to create the output.

Notice I used "-shave 1" before the "-border 1" so the borders don't change the dimensions. That way all the various sized images still line up when appended.

**Comment by Trailboy** [comment]
GeeMack, Thank you so much for helping me with this.  I will experiment with your code and advice shortly.  

I have been experimenting over the last hour with my own code, and have come up with this which seems to work - but given my inexperience, might be incorrect.  Is there any obvious reason the following is wrong?  

```
magick ^
	( %1 -rotate "-90>" -resize 600 -gravity NorthWest -extent 1200x1800 ) ^
	( %1 -rotate "-90>" -resize 600 -write mpr:small ) ^
	-gravity NorthWest -geometry +600+0 -composite ^
	mpr:small -gravity NorthWest -geometry +1200+0 -composite ^
	( %1 -rotate "-90>" -resize 300 -gravity NorthWest ) ^
	( %1 -rotate "-90>" -resize 300 -write mpr:small ) ^
	-gravity NorthWest -geometry +0+900 -composite ^
	mpr:small -gravity NorthWest -geometry +300+900 -composite ^
	mpr:small -gravity NorthWest -geometry +600+900 -composite ^
	mpr:small -gravity NorthWest -geometry +900+900 -composite ^
	mpr:small -gravity NorthWest -geometry +0+1350 -composite ^
	mpr:small -gravity NorthWest -geometry +300+1350 -composite ^
	mpr:small -gravity NorthWest -geometry +600+1350 -composite ^
	mpr:small -gravity NorthWest -geometry +900+1350 -composite ^
	"%~dp1\%~n1_KeyringsMagnets.jpg"
```

which gives me:

![Imagemagick-3](https://user-images.githubusercontent.com/64076927/99150335-b3930b00-268b-11eb-9d83-b617b10289f0.JPG)


Given more time, I might be able to get much more complicated layouts, such as multiple 12 inch, 9 inch and 6 inch prints on one 18 inch image.

**Reply by GeeMack** [reply]
With just a quick look at your command I'd make a couple suggestions. First, you only need to set the gravity once if everything after uses the same setting. You can put "-gravity northwest" right at the very beginning of the command.

Also I think it would be more efficient to read the original image into the command one time at the start, write it to a memory register "mpr:something", then delete it. After that use "mpr:something" instead of "%1" to avoid reading from disk multiple times.

    magick %1 ^
        -write mpr:input ^
        +delete ^
        ( mpr:input ... ) ^
        ...

Generally it looks like your command is more complicated than it needs to be. I don't know for sure, but it seems using multiple composites instead of appends might be less efficient.

**Reply by GeeMack** [reply]
You may be able to simplify that command quite a lot with something like this...

    magick %1 ^
        -rotate "-90>" ^
        -gravity center ^
        -extent 2:3 ^
        +duplicate ^
        +append ^
        ( +clone ^
            -resize 50%% ^
            +duplicate ^
            +append ^
            +duplicate ^
            -append ) ^
        -append ^
        -resize 1200x ^
        "%~n1_Layout1.jpg"

That reads the input image, checks it for landscape or portrait and rotates it if needed, and crops it to exactly 2:3 aspect ratio. Then it duplicates it and appends those two horizontally. Next, inside the parentheses it resizes a clone to 50%, then duplicates and appends them to make the 4 by 2 grid of small ones. After the parentheses it appends the grid of small ones below the pair of larger ones, resizes that to your finished output dimensions, and produces a result like your command.

**Comment by Trailboy** [comment]
Thanks GeeMack, your code does seem to simplify the process.  

I did progress a little bit on the weekend with my multiple layouts, and came up with the following.  

I think my code is very cludgy.  For example, I need to repeat the -border every time and I'm not sure why, but I have to have borders because I sometimes photograph portraits on a pure white background.   It does, however, give me the print layout/s I want.  

Any help cleaning this code up would be very appreciated.  

Thanks in advance.  

---
Layout 1 - one 9x6 inch sized print (1800x2700 pixels) with a tiled layout of one 6x4, three 3x2 (fridge magnet sized) and eight 1.5x1 (keyrings sized).

![Imagemagick-5](https://user-images.githubusercontent.com/64076927/99380996-97c57a00-28c2-11eb-93b3-fa2a2a4250fb.JPG)

```
magick %1 ^
	-write mpr:input ^
	+delete ^
	( mpr:input -rotate "-90>" -resize 1200 -shave 1 -bordercolor gray -border 1 -gravity NorthWest -extent 1800x2700 ) ^
	( mpr:input -rotate "-90>" -resize 1200 -shave 1 -bordercolor gray -border 1 -write mpr:small ) ^
	-geometry +0+0 -composite ^
	( mpr:input -rotate "-90>" -resize 600 -shave 1 -bordercolor gray -border 1 ) ^
	( mpr:input -rotate "-90>" -resize 600 -shave 1 -bordercolor gray -border 1 -write mpr:small ) ^
	-geometry +0+1800 -composite ^
	mpr:small  -geometry +600+1800 -composite ^
	mpr:small  -geometry +1200+1800 -composite ^
	( mpr:input -rotate "-90>" -resize 300 -shave 1 -bordercolor gray -border 1 ) ^
	( mpr:input -rotate "-90>" -resize 300 -shave 1 -bordercolor gray -border 1 -write mpr:small ) ^
	-geometry +1200+0 -composite ^
	mpr:small -geometry +1500+0 -composite ^
	mpr:small -geometry +1200+450 -composite ^
	mpr:small -geometry +1500+450 -composite ^
	mpr:small -geometry +1200+900 -composite ^
	mpr:small -geometry +1500+900 -composite ^
	mpr:small -geometry +1200+1350 -composite ^
	mpr:small -geometry +1500+1350 -composite ^
	"%~dp1\%~n1_PackEventCombo-9x6.jpg"
```
---

Layout 2 - one 12x8 inch sized print (2400x3600 pixels) with a tiled layout of three 6x4, two 3x2 and eight 1.5x1.

![Imagemagick-6](https://user-images.githubusercontent.com/64076927/99381714-597c8a80-28c3-11eb-9cb9-eefdaaa1379e.JPG)

```
magick %1 ^
	-write mpr:input ^
	+delete ^
	( mpr:input -rotate "-90>" -resize 1200 -shave 1 -bordercolor gray -border 1 -gravity NorthWest -extent 2400x3600 ) ^
	( mpr:input -rotate "-90>" -resize 1200 -shave 1 -bordercolor gray -border 1 -write mpr:small ) ^
	-geometry +1200+0 -composite ^
	mpr:small  -geometry +1200+0 -composite ^
	mpr:small  -geometry +0+1800 -composite ^
	( mpr:input -rotate "-90>" -resize 600 -shave 1 -bordercolor gray -border 1 ) ^
	( mpr:input -rotate "-90>" -resize 600 -shave 1 -bordercolor gray -border 1 -write mpr:small ) ^
	-geometry +1200+2700 -composite ^
	mpr:small  -geometry +1800+2700 -composite ^
	( mpr:input -rotate "-90>" -resize 300 -shave 1 -bordercolor gray -border 1 ) ^
	( mpr:input -rotate "-90>" -resize 300 -shave 1 -bordercolor gray -border 1 -write mpr:small ) ^
	-geometry +1200+1800 -composite ^
	mpr:small -geometry +1500+1800 -composite ^
	mpr:small -geometry +1800+1800 -composite ^
	mpr:small -geometry +2100+1800 -composite ^
	mpr:small -geometry +1200+2250 -composite ^
	mpr:small -geometry +1500+2250 -composite ^
	mpr:small -geometry +1800+2250 -composite ^
	mpr:small -geometry +2100+2250 -composite ^
    -border 1 ^
	"%~dp1\%~n1_PackG-12x8.jpg"

```

**Reply by GeeMack** [reply]
I wouldn't use that approach of using multiple composites for this task. Here are a couple commands that create the same results as your above commands, but these are simpler to read and write, and they're much faster.

This command makes your first layout above...

    magick %1 -rotate "-90>" -extent 2:3 -resize 1200x ^
       ( +clone -resize 50%% ) ( +clone -resize 50%% ) ^
       -bordercolor gray -shave 1 -border 1 ^
       -write mpr:imgs -delete 0--1 ^
       mpr:imgs[0] ^
       ( mpr:imgs[2] -duplicate 3 -append +duplicate +append ) ^
       +append ^
       ( mpr:imgs[1] -duplicate 2 +append ) ^
       -append ^
          "%~dp1\%~n1_PackEventCombo-9x6.jpg"

That starts by orienting and sizing the input image, making a half size copy and a quarter size copy, adding a border to each image, then writing all three of them to a memory register "mpr:imgs". As the command continues you can access those memory images with "mpr:imgs[0]", "mpr:imgs[1]", and "mpr:imgs[2]".

After the three sizes are created and bordered, the sub-parts get assembled inside sets of parentheses, and those parts are assembled outside the parentheses.

This command makes your second layout. It works exactly the same way but with a different order and arrangement of appending the parts...

    magick %1 -rotate "-90>" -extent 2:3 -resize 1200x ^
       ( +clone -resize 50%% ) ( +clone -resize 50%% ) ^
       -bordercolor gray -shave 1 -border 1 ^
       -write mpr:imgs -delete 0--1 ^
       ( mpr:imgs[2] -duplicate 3 +append +duplicate -append ) ^
       ( mpr:imgs[1] +duplicate +append ) ^
       -append ^
       mpr:imgs[0] ^
       +swap +append ^
       ( mpr:imgs[0] +duplicate +append ) ^
       +swap -append ^
          "%~dp1\%~n1_PackG-12x8.jpg"

Note the borders are added to each size image all at once and before the assembling them so all the borders are the same thickness.

Also note the use of "+swap" to trade the order of the parts before appending.

I've compared both these commands with the ones you provided, and although speed may not be a priority, this approach is almost twice as fast.

**Comment by Trailboy** [comment]
Hi Geemack.  I can only thank you for the valuable time you are giving me.  It is greatly appreciated.

I was going to attempt to dissect and understand your code tonight, but it doesn't seem to work for me - your output has -Geemack appended.  

Any reason why I'm getting the black?  

![Imagemagick-7](https://user-images.githubusercontent.com/64076927/99445017-e4d23c00-2914-11eb-9d4b-1d518412bc8f.JPG)

I think I'm running one of the latest versions of IM7.  

![Magick Version](https://user-images.githubusercontent.com/64076927/99445124-121eea00-2915-11eb-81f5-3cbb634c5ffa.JPG)

**Reply by GeeMack** [reply]
With a quick look I found an error in my first command with this line...

    ( mpr:imgs[1] -duplicate 2 +append ) ^

That line creates the bottom row, the row with three images. It wasn't duplicating enough. I edited the command to correct that.

**Comment by Trailboy** [comment]
Sorry Geemack, I'm still getting the same result as earlier.  It appears to only be writing the first image.

![ImageMagick9](https://user-images.githubusercontent.com/64076927/99502889-9acd7280-2975-11eb-81a1-bd3578a3c828.JPG)

**Reply by GeeMack** [reply]
I use the same version of IM "ImageMagick 7.0.10-38 Q16 x64 HDRI" on Windows 10, and the results are correct. We can test a command to see where it fails.

Here is the command I used for the 9x6 layout, same as posted above, but note how I added a "-write test.png" line to have it write some intermediate files to disk. This command should write three images, "test-0.png", "test-1.png", and "test-2.png". See if the second and third images are resized to 50% and 25%. If they look correct at that point, move the "-write test.png" to right after adding the borders and check those results. It should be the same three images with gray borders. Then continue to test by moving that write operation down below each next operation. and check the results at each step.

    magick %1 -rotate "-90>" -extent 2:3 -resize 1200x ^
       ( +clone -resize 50%% ) ( +clone -resize 50%% ) ^
                                                              -write test.png ^
       -bordercolor gray -shave 1 -border 1 ^
       -write mpr:imgs -delete 0--1 ^
       mpr:imgs[0] ^
       ( mpr:imgs[2] -duplicate 3 -append +duplicate +append ) ^
       +append ^
       ( mpr:imgs[1] -duplicate 2 +append ) ^
       -append ^
          "%~dp1\%~n1_PackEventCombo-9x6.jpg"

If you can find where the command creates the black images, we can probably figure out why and easily fix it.

**Comment by Trailboy** [comment]
Really sorry to be using your time like this, but using the exact above code, the files are created black with the '-write test.png' at line 3.  They are created at the correct dimensions however.  

I do use Windows7 for this machine, might this be the cause?  

![ImageMagick11](https://user-images.githubusercontent.com/64076927/99552165-98d6d400-29b4-11eb-9c0f-122d414a5b71.JPG)

![ImageMagick10](https://user-images.githubusercontent.com/64076927/99552172-9aa09780-29b4-11eb-88a7-40a79954acb8.JPG)

**Reply by GeeMack** [reply]
No problem. I'm glad to help. I don't have any idea why the clones turn black. May be a bug that acts up on Windows 7, I dunno. We could try a couple other ways to make those clones. I think first I'd try to resize that input image then write it to a memory register "mpr:test", then call it back into the command with "mpr:test" instead of a clone. 

    magick %1 -rotate "-90>" -extent 2:3 -resize 1200x -write mpr:test ^
       ( mpr:test -resize 50%% ) ( mpr:test -resize 25%% ) ^
                                                              -write test.png ^
       -bordercolor gray -shave 1 -border 1 ^
       -write mpr:imgs -delete 0--1 ^
       mpr:imgs[0] ^
       ( mpr:imgs[2] -duplicate 3 -append +duplicate +append ) ^
       +append ^
       ( mpr:imgs[1] -duplicate 2 +append ) ^
       -append ^
          "%~dp1\%~n1_PackEventCombo-9x6.jpg"

See if those test images still turn black.

**Comment by Trailboy** [comment]
Thanks Geemack, a working Pack creation system with Imagemagick would save me a huge amount of time if we can get it working.  

With the above code, all files are being created to the correct dimensions, but the middle sized image seems only to be creating the first  few rows of pixels and then seems to give up.  Would this point more a bug?

See the (rather large, sorry) attachment.  

![ImageMagick13](https://user-images.githubusercontent.com/64076927/99679544-f54cf880-2a73-11eb-81af-bf475b52ee7b.JPG)

**Comment by snibgo** [comment]
GeeMack's command above...
```
magick %1 -rotate "-90>" -extent 2:3 -resize 1200x ^
   ( +clone -resize 50%% ) ( +clone -resize 50%% ) ^
                                                          -write test.png ^
   -bordercolor gray -shave 1 -border 1 ^
   -write mpr:imgs -delete 0--1 ^
   mpr:imgs[0] ^
   ( mpr:imgs[2] -duplicate 3 -append +duplicate +append ) ^
   +append ^
   ( mpr:imgs[1] -duplicate 2 +append ) ^
   -append ^
      "%~dp1\%~n1_PackEventCombo-9x6.jpg"
```
... works fine for me, in a BAT script, using my usual toes.png and other images, with IM 7.0.8-64 on Windows 8.1.

I can't reproduce the black problem, eg with an input that has a canvas offset. But there may be some issue with the input file.

Here is toes.jpg:
![toes](https://user-images.githubusercontent.com/33812368/99688125-5a591c00-2a7d-11eb-9890-85fbfda55d06.jpg)

@Trailboy: Please download toes.jpg, and paste this into a BAT file, and run it:
```
magick ^
  toes.jpg ^
  -rotate "-90>" +write x0.png ^
  -extent 2:3 +write x1.png ^
  -resize 1200x +write x2.png ^
  ( +clone -resize 50%% ) ^
  ( +clone -resize 50%% ) ^
  -write x3.png ^
  -bordercolor gray -shave 1 -border 1 +write x4.png ^
  -write mpr:imgs -delete 0--1 ^
  mpr:imgs[0] ^
  ( mpr:imgs[2] -duplicate 3 -append +duplicate +append +write x5.png ) ^
  +append ^
  +write x6.png ^
  ( mpr:imgs[1] -duplicate 2 +append +write x7.png ) ^
  -append ^
  toes-9x6.jpg
```
This creates many debugging images. Which of these is wrong? This should isolate the problem.

**Comment by Trailboy** [comment]
Hi Snibgo. Thanks for contributing.  

I get the same problems as earlier.  Only the first image is being written properly, which must point to a problem with my installation I suppose?  

![ImageMagick14](https://user-images.githubusercontent.com/64076927/99709561-41f5fb00-2a97-11eb-86e6-c0548604b28e.JPG)

**Comment by snibgo** [comment]
To try and help you, I suggested you run a command that creates x0.png etc. But you haven't. Why not?

**Reply by Trailboy** [reply]
Snibgo, please accept my apologies, it's been a long day in work and I really should engage the brain before posting.  

Does this shed any light?

![Image Magick 15](https://user-images.githubusercontent.com/64076927/99717749-bdf54080-2aa1-11eb-9dfe-6649ab22e1e2.JPG)

**Comment by snibgo** [comment]
That helps.

x0.png, x1.png and x2.png look okay. The first wrong output is x3-2.png. This is the second "+clone -resize 50%%". The first of those seems okay.

The top of x3-2.png looks okay, but the rest is black. This suggests an OpenMP (or possibly OpenCL) problem. Please re-run the test, adding a line at the top:
```
magick ^
  -limit thread 1 ^
  toes.jpg ^
  -rotate "-90>" +write x0.png ^
  ... etc
```

**Reply by Trailboy** [reply]
Thanks for this. 

The  -limit thread 1 does not seem to have had any effect.  

![ImageMagick 18](https://user-images.githubusercontent.com/64076927/99722522-acfbfd80-2aa8-11eb-9113-2b527fbaded1.JPG)

**Comment by snibgo** [comment]
Let's try disabling OpenCL. In your BAT file, add a command before the magick command.
```
set MAGICK_OCL_DEVICE=off

magick ^
  -limit thread 1 ^
  toes.jpg ^
  -rotate "-90>" +write x0.png ^
  ... etc
```
What happens now?

**Comment by Trailboy** [comment]
It appears that that OpenCL is my problem.  Appears to work perfectly now.  What would your advice be?

![ImageMagick 19](https://user-images.githubusercontent.com/64076927/99727320-b046b780-2aaf-11eb-824b-79d833a99fa6.JPG)

**Comment by snibgo** [comment]
I suggest removing the "-limit thread 1" line and re-run the test. Try it a few times. If it works reliably, then we can assume OpenCL is the problem.

Bingo! Hoorah!

IM uses OpenCL to use the computer's GPU for some operations. I guess your computer is quite old (as it runs Windows 7), and perhaps OpenCL doesn't properly support your GPU.

Anyhow, if this cures the problem, I suggest you set "MAGICK_OCL_DEVICE=off" system-wide. On Windows 8.1, this is at Control panel => System => Advanced system settings => Advanced tab => Environment variables. Create an environment variable called MAGICK_OCL_DEVICE and set its value to off. Windows 7 may be slightly different.

**Comment by Trailboy** [comment]
Thanks Geemack and Snibgo for all your help with this.

That OpenCL does appear to be the problem. Geemack's earlier examples now work perfectly with that 'set MAGICK_OCL_DEVICE=off' on line 1, and you are right, they are far faster than my code.  So many thanks.  Now I have to figure out how they work so that I can apply it to other layouts that I need. 

The machines aren't too old but are all NUCs (hence Intel HD drivers), but I need to run Windows 7 on them all due to printers that never received Win8/10 drivers.  They are normally offline when operating, just not very much at the mo (thanks to Covid).  

Thanks again both; I'll return to annoy the hell out of you again when I can't figure out the earlier code.

**Comment by snibgo** [comment]
I should add: there is driver software that sits between OpenCL and the GPU. Perhaps your problem would be fixed by an update to that driver software. I don't know how to check driver software versions or how to update it, but that is something you might investigate, if you want.

**Comment by Trailboy** [comment]
Hi again.  

Sorry to have to come back and ask for help yet again.  Really hope you don't mind.  

Now the academic year is coming to an end, I have more time to experiment with the invaluable help previously given above to create more complex print layouts, but have hit snags with two layouts that I can't resolve at all.

As you can see, I have a problem with layouts PackA and PackD.  

![Layout1](https://user-images.githubusercontent.com/64076927/122745417-e2f0a280-d280-11eb-8442-081e638c417b.JPG)

Pack A code:

```
set MAGICK_OCL_DEVICE=off

magick %1 -rotate "-90>" -gravity northwest -extent 2:3 -quality 100 -resize 3600x3600 ^
	( +clone -resize 2700x2700 ) ( +clone -resize 1800x1800 ) ( +clone -resize 900x900 ) ( +clone -resize 450x450 )^
	-bordercolor gray -shave 3 -border 3 ^
	-write mpr:imgs -delete 0--1 ^
	( mpr:imgs[3] -rotate -180 -duplicate 1 -append ) ^
	( mpr:imgs[2] -rotate -180 -duplicate 1 +append ) ^
	+swap +append ^
	( mpr:imgs[4] -duplicate 3 -append ) ^
	( mpr:imgs[1] -rotate -90 -append ) ^
	+swap -append ^
	mpr:imgs[0]  ^
	+swap +append ^
	"%~dp1\%~n1_PackA-18x12.jpg"
```

gives me:

![Layout2](https://user-images.githubusercontent.com/64076927/122746639-1ed83780-d282-11eb-8307-b62d635f94b4.JPG)

and I cannot get the 4 small thumbnail images overhanging the bottom to align in the gap to the right of the bottom right image.  

Pack D code:

```
set MAGICK_OCL_DEVICE=off

magick %1 -rotate "-90>" -gravity northwest -extent 2:3 -quality 100 -resize 2700x2700 -rotate 270 ^
	( +clone -resize 1800x1800 ) ( +clone -resize 900x900 ) ( +clone -resize 450x450 ) ^
	-bordercolor gray -shave 3 -border 3 ^
	-write mpr:imgs -delete 0--1 ^
	( mpr:imgs[1] -rotate -90 -duplicate 1 +append ) ^
	( mpr:imgs[1] -rotate -270 -append ) ^
	( mpr:imgs[2] -rotate -270 -duplicate 1 +append ) ^
	( mpr:imgs[3] -rotate -270 -duplicate 3 +append +duplicate -append ) ^
	-append ^
	( mpr:imgs[0] -rotate -180 -duplicate 1 -append ) ^
	+swap +append ^
	"%~dp1\%~n1_PackD-18x12.jpg"
```

give me 

![Layout3](https://user-images.githubusercontent.com/64076927/122747302-cf463b80-d282-11eb-8274-5238108537bd.JPG)

i.e., a similar problem as earlier, with bottom hanging thumbnail images that I can't get to align on the gap.  

Any help in where I am going wrong would be greatly appreciated.  

I have to admit, that my current layouts are more by experimental luck than correctly  following the logic of +append and -append multiple times for such complex layouts.  

Thanks in advance to anybody that can help.

**Reply by fmw42** [reply]
If you need help debugging your code, please always provide the input image (at full resolution and by itself). The best way to debug code is to add +write tmpX.png for X=1, 2, 3 ... inside each parenthesis right at the end and also after each step non-parenthesis operator to make sure the results are as you expected. We can only do this to help if we have your input image.

**Reply by fmw42** [reply]
The first result is going wrong at

( mpr:imgs[4] -duplicate 3 -append ) ^
( mpr:imgs[1] -rotate -90 -append ) ^
+swap -append ^

your are appending vertically. You need to append horizontally [1] and [4] without appending the image above it. So you likely need another nest of parenthesis

( ^
( mpr:imgs[4] -duplicate 3 -append ) ^
( mpr:imgs[1] -rotate -90 -append ) ^
+swap -append  ) ^

**Reply by fmw42** [reply]
Try this

![monet2](https://user-images.githubusercontent.com/13458218/122803707-d158cc00-d27b-11eb-949e-83b9e5e961c0.jpg)


```
magick monet2.jpg ^
-rotate "-90>" ^
-gravity northwest -extent 2:3 ^
-resize 3600x3600 ^
( +clone -resize 2700x2700 ) ^
( +clone -resize 1800x1800 ) ^
( +clone -resize 900x900 ) ^
( +clone -resize 450x450 ) ^
-bordercolor gray -shave 3 -border 3 ^
-write mpr:imgs -delete 0--1 ^
( ^
( mpr:imgs[3] -rotate -180 -duplicate 1 -append ) ^
( mpr:imgs[2] -rotate -180 -duplicate 1 +append ) ^
+swap +append ^
) ^
( ^
( mpr:imgs[4] -duplicate 3 -append ) ^
( mpr:imgs[1] -rotate -90 -append ) ^
+swap +append ^
) ^
-append ^
mpr:imgs[0]  ^
+swap +append ^
-quality 100 ^
result.jpg

```

![result](https://user-images.githubusercontent.com/13458218/122803747-de75bb00-d27b-11eb-99a0-7a188084b8c8.jpg)

**Comment by Trailboy** [comment]
Hi fmw42

Sorry, I should have included it.  

Drive link: https://drive.google.com/drive/folders/12o4qifV2hWMfjJgrgwM1wqcIrWwD4ngu?usp=sharing

Will try and have a go with your recommendations.

**Reply by fmw42** [reply]
Unix syntax equivalent:

![Sample](https://user-images.githubusercontent.com/13458218/122809979-a70b0c80-d283-11eb-89a5-a18479af74b5.jpg)


```
magick sample.jpg \
-rotate "-90>" \
-gravity northwest -extent 2:3 \
-resize 3600x3600 \
\( +clone -resize 2700x2700 \) \
\( +clone -resize 1800x1800 \) \
\( +clone -resize 900x900 \) \
\( +clone -resize 450x450 \) \
-bordercolor gray -shave 3 -border 3 \
-write mpr:imgs -delete 0--1 \
\( \
\( mpr:imgs[3] -rotate -180 -duplicate 1 -append \) \
\( mpr:imgs[2] -rotate -180 -duplicate 1 +append \) \
+swap +append \
\) \
\( \
\( mpr:imgs[4] -duplicate 3 -append \) \
\( mpr:imgs[1] -rotate -90 -append \) \
+swap +append \
\) \
-append \
mpr:imgs[0]  \
+swap +append \
-quality 100 \
result.jpg

```

![result](https://user-images.githubusercontent.com/13458218/122810014-b12d0b00-d283-11eb-9ba6-bba9d33e68fe.jpg)

**Reply by Trailboy** [reply]
Thanks, that's much better.  

Will try and work out the logic in your (much cleaner) code and apply it to my 'PackD' code.

I can already see a mistake in line 1 of my code for PackD with an incorrect initial -rotate 270.  Don't know where that came from.

**Comment by Trailboy** [comment]
Thanks so much Fred for the direction above - especially using nested parentheses.  

I think I have managed to correctly create my PackD layout now (with a slight layout change) using the following code, unless you think it needs to be improved in any way??? :

```
set MAGICK_OCL_DEVICE=off

magick %1 ^
	-rotate "-90>" ^
	-gravity northwest ^
	-extent 2:3 ^
	-resize 2700x2700 ^
		( +clone -resize 1800x1800 ) ^
		( +clone -resize 900x900 ) ^
		( +clone -resize 450x450 ) ^
	-bordercolor gray -shave 3 -border 3 ^
	-write mpr:imgs -delete 0--1 ^
	( ^
		( mpr:imgs[2] -rotate -180 -duplicate 3 -append ) ^
		( mpr:imgs[3] -rotate -180 -duplicate 7 -append ) ^
		+swap +append ^
	) ^
	( ^
		( mpr:imgs[1] -rotate -90 -duplicate 2 -append ) ^
	) ^
	+append ^
		( mpr:imgs[0] -duplicate 1 -rotate -270 -append ) ^
	+swap +append ^
	-quality 100 ^
	"%~dp1\%~n1_PackD-18x12.jpg"
```
gives: 

![Sample_PackD-18x12](https://user-images.githubusercontent.com/64076927/122890191-5e168f00-d33b-11eb-9e8b-e844a4a1a1f9.jpg)

**Reply by fmw42** [reply]
Looks reasonable. The only comment would be to work left to right to avoid all the swaps.

**Comment by snibgo** [comment]
I suggest that "-append" and "+append" should _always_ be followed by "+repage", unless there is a good reason otherwise.

The reason is that after appending images, the first image is used for the canvas size metadata. For example:
```
magick -size 10x15 xc: xc: +append info:

xc: XC 20x15 10x15+0+0 16-bit sRGB 0.000u 0:00.003
```
Note the result is correctly 20x15 pixels, but on a canvas that is only 10x15. This can cause problems.

In this particular case, nothing uses the canvas size, and JPG files don't store them, so this makes no difference. But I think it is good practice to always use "+repage".

**Comment by Trailboy** [comment]
Thanks Snibgo and Fred. 

One more query.  I am trying to incorporate the commands into a .bat file to run over all jpegs within a folder, but would like to retain my line breaks for easier understanding.  I've researched and apparently the Caret ^ is allowed within Windows .bat files, but mine isn't working.  I suspect it is to do with the required double percentage symbols, but unsure myself.    

This works:

`for %%F in (*.jpg) do magick %%F -rotate "-90>" -gravity northwest -extent 2:3 -quality 100 -resize 2700x2700 ( +clone -resize 1800x1800 ) ( +clone -resize 450x450 ) -bordercolor gray -shave 3 -border 3 -write mpr:imgs -delete 0--1 ( mpr:imgs[1] -rotate -90 -duplicate 1 -append ) ( mpr:imgs[2] -rotate -90 -duplicate 3 +append ) -append mpr:imgs[0] +swap +append %%~nF_PackF-12x9.jpg`

but this doesn't:

```
for %%F in (*.jpg) do magick %%F ^ 
	-rotate "-90>" ^
	-gravity northwest ^
	-extent 2:3 ^
	-quality 100 ^
	-resize 2700x2700 ^
		( +clone -resize 1800x1800 ) ( +clone -resize 450x450 ) ^
	-bordercolor gray -shave 3 -border 3 ^
	-write mpr:imgs -delete 0--1 ^
		( mpr:imgs[1] -rotate -90 -duplicate 1 -append ) ^
		( mpr:imgs[2] -rotate -90 -duplicate 3 +append ) ^
	-append ^
		mpr:imgs[0] ^
	+swap +append ^
	%%~nF_PackF-12x9.jpg
```

Ideas on how to fix, or confirmation that it isn't possible, would be greatly appreciated.

**Comment by snibgo** [comment]
When the caret is the last character on the line, it means the command continues on the next line.

When the caret is not the last character, it means the next character should be used literally.

And that is your problem: the carets should be the last characters.

**Reply by Trailboy** [reply]
An extra space after the first caret was all it took!?! 

Thanks Snibgo!

**Reply by snibgo** [reply]
Yes. Caret and space means "treat the caret and space as a literal space", which is fairly pointless, and the command ends there, calling "magick" with too few arguments. The interpreter would read the next line as a command, and probably complained that "-rotate" was not recognized as a valid command.
